presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -440,18 +441,23 @@ public PlanWithProperties visitTopNRowNumber(TopNRowNumberNode node, Context con
         @Override
         public PlanWithProperties visitTopN(TopNNode node, Context context)
         {
-            PlanWithProperties child = planChild(node, context.withPreferredProperties(PreferredProperties.any()));
-
-            if (!child.getProperties().isSingleNode()) {
-                child = withDerivedProperties(
-                        new TopNNode(idAllocator.getNextId(), child.getNode(), node.getCount(), node.getOrderBy(), node.getOrderings(), TopNNode.Step.PARTIAL),
-                        child.getProperties());
-
-                child = withDerivedProperties(
-                        gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
-                        child.getProperties());
+            PlanWithProperties child;
+            switch (node.getStep()) {
+                case SINGLE:
```

Comment:
I don't see a drawback of supporting `SINGLE` as well. It gives us more flexibility, in case of some refactor or especially in tests (where programmers are handpicking optimizers).

However yes, with iterative optimizer disabled we need this.